#!/bin/bash
#7_invoke_get_homologues
#
#Author: Abelardo Aguilar Camara
#
ANYTAXON=$1 # $1 = anytaxon aka viral family (e.g. Circoviridae)
LOWLIM=$2 # $2 = LOWLIM which is the minimum number of .gbk within a cluster for it to be considered as pangenomic input cluster (not less than 3)
NUMTHREADS=$3 # $3 = NUMTHREADS, number of threads for local blast to run
MINCOVER=$4 # $4 = MINCOVER, minimum % of coverage in BLAST pairwise alignments
MAX_EVALUE=$5 # $5 = MAX_EVALUE, maximum E-value from BLAST (max accepted is 0.01) 
STARTTIME=`date +"%Y-%m-%d %T"`
ENDTIME=`date +"%d%m%Y_%H%M"` # Variable to print in folder name

#LOWLIM MUST BE GREATER THAN 3
while [[ $LOWLIM -lt 3 ]] ; do
  echo -ne "\nLOWLIM must be equal or greater than 3\n\n"
  read -p "type the minimum number of .gbk within a cluster for it to be considered as pangenomic input cluster (not less than 3):" LOWLIM
done

#LIST FILES TO BE USED AS PANGENOMIC INPUTS
echo -ne "\nFollowing directories will be used for get_homologues:\n"
find ../results/Pangenomic_input_clusters/ -mindepth 2 -maxdepth 2 -type d -exec bash -c "echo -ne '{} '; ls '{}' | wc -l" \; |
 awk -v LOWLIM=$LOWLIM '$NF>=LOWLIM' | grep $ANYTAXON

#PERFORM PANGENOMIC ANALYSIS WITH GETHOMS
find ../results/Pangenomic_input_clusters/ -mindepth 2 -maxdepth 2 -type d -exec bash -c "echo -ne '{} '; ls '{}' | wc -l" \; |
 awk -v LOWLIM=$LOWLIM '$NF>=LOWLIM' | grep $ANYTAXON | cut -d' ' -f1 |
  while read paninput ; do get_homologues.pl -d $paninput -D -G -n $NUMTHREADS -t 0 -z -C $MINCOVER -E $MAX_EVALUE; get_homologues.pl -d $paninput -D -M -n $NUMTHREADS -t 0 -z -C $MINCOVER -E $MAX_EVALUE; done

#MOVE FOLDERS GENERATED BY GET_HOMOLOGUES
mkdir -p ../results/Get_homologues
mkdir -p ../results/Get_homologues/$ANYTAXON
find . -maxdepth 1 -newermt "$STARTTIME" | grep "homologues" | while read file ; do mv $file "${file}_${ENDTIME}"; done
find . -maxdepth 1 -newermt "$STARTTIME" | grep "homologues" | while read file ; do mv $file ../results/Get_homologues/$ANYTAXON; done
